src/test/kotlin/com/coder/gateway/sdk/BaseCoderRestClientTest.kt

```diff
@@ -39,47 +52,104 @@ class BaseCoderRestClientTest {
+        var matches = resourceEndpoint.find(exchange.requestURI.path)
+        if (exchange.requestMethod == "GET" && matches != null) {
+            val templateVersionId = UUID.fromString(matches.destructured.toList()[0])
+            spy?.invoke(SpyAction.GET_RESOURCES, templateVersionId)
+            val ws = workspaces.firstOrNull { it.workspace.latestBuild.templateVersionID == templateVersionId }
+            if (ws != null) {
+                return Pair(HttpURLConnection.HTTP_OK, toJson(ws.resources))
+            }
+        }
+
+        matches = templateEndpoint.find(exchange.requestURI.path)
+        if (exchange.requestMethod == "GET" && matches != null) {
+            val templateId = UUID.fromString(matches.destructured.toList()[0])
+            spy?.invoke(SpyAction.GET_TEMPLATE, templateId)
+            val template = templates.firstOrNull { it.id == templateId }
+            if (template != null) {
+                return Pair(HttpURLConnection.HTTP_OK, toJson(template))
+            }
+        }
+
+        matches = buildEndpoint.find(exchange.requestURI.path)
+        if (exchange.requestMethod == "POST" && matches != null) {
+            val workspaceId = UUID.fromString(matches.destructured.toList()[0])
+            val json = Gson().fromJson(InputStreamReader(exchange.requestBody), CreateWorkspaceBuildRequest::class.java)
+            if (json.templateVersionID != null) {
+                spy?.invoke(SpyAction.UPDATE_WORKSPACE, workspaceId)
+            } else {
+                when (json.transition) {
+                    WorkspaceTransition.START -> spy?.invoke(SpyAction.START_WORKSPACE, workspaceId)
+                    WorkspaceTransition.STOP -> spy?.invoke(SpyAction.STOP_WORKSPACE, workspaceId)
+                    WorkspaceTransition.DELETE -> Unit
+                }
+            }
+            val ws = workspaces.firstOrNull { it.workspace.id == workspaceId }
+            if (ws != null) {
+                return Pair(HttpURLConnection.HTTP_CREATED, toJson(ws.workspace))
+            }
+        }
+
+        when (exchange.requestURI.path) {
+            "/api/v2/workspaces" -> {
+                spy?.invoke(SpyAction.GET_WORKSPACES, null)
+                return Pair(HttpsURLConnection.HTTP_OK, toJson(WorkspacesResponse(workspaces.map{ it.workspace }, workspaces.size)))
+            }
+            "/api/v2/users/me" -> {
+                spy?.invoke(SpyAction.GET_ME, null)
+                val user = User(
+                    UUID.randomUUID(),
+                    "tester",
+                    "tester@example.com",
+                    Instant.now(),
+                    Instant.now(),
+                    UserStatus.ACTIVE,
+                    listOf(),
+                    listOf(),
+                    "",
+                )
+                return Pair(HttpsURLConnection.HTTP_OK, toJson(user))
+            }
+        }
+        return Pair(HttpsURLConnection.HTTP_NOT_FOUND, "not found")
+    }
```

Comment:
Down the line I could imagine other tests having to specify different behaviours for these endpoints. In that case, it'd be nicer to have those behaviours in the actual test definition itself as opposed to all being stuffed in here.

Reply:
Holy smokes that would be so much better, and I can get rid of all this spy nonsense.  I will refactor.